[Workers AI, Vectorize] Add a new tutorial about building product recommendation API

hideokamoto-stripe opened the pull request:
Here is a part of the content of Cloudflare Developer Spotlight program. We're going to introduce the use case based tutorial using Cloudflare AI stacks ( Workers AI and Vectorize ).

If you have any question, feedback, or concern, please feel free let us know.

Thanks.

yusukebe commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
````diff
@@ -0,0 +1,847 @@
+
+In this step, we'll first implement the process of converting externally sent data into text data. This is necessary because the information to be converted into vector data is in text form. If you want to include product names, descriptions, and metadata as search targets, add the following processing.
+
+```diff
````
You can use a `highlight` feature instead of the `diff` like the following:

```markdown
```ts
---
filename: src/index.ts
highlight: [3, 4, 5, 6, 7, 8, 9]
---
if (event.type === 'product.created') {
  const product = event.data.object
  const productData = [
    `## ${product.name}`,
    product.description,
    '### metadata',
    Object.entries(product.metadata).map(([key, value]) => `- ${key}: ${value}`).join('\n')
  ].join('\n')
  console.log(productData)
}
```

In this way, you can highlight the specified part while keeping the TypeScript syntax highlighting. I think you should use a highlight for other places for the same purpose, where include the `wrangler.toml`.

yusukebe commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
````diff
@@ -0,0 +1,847 @@
+This method realizes a text generation mechanism called Retrieval Augmented Generation (RAG) using Cloudflare. The bindings and other preparations are already completed, so let's add the API.
+
+```ts
+
````
This line should be removed.

yusukebe commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,847 @@
+    )
+    if (event.type === 'product.created') {
+      const product = event.data.object
+      console.log(product)
```
To dump and format it properly, the below would be better.

```ts
console.log(JSON.stringify(product, undefined, 2))
```

yusukebe commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
````diff
@@ -0,0 +1,847 @@
+Let's run this API. Use a product ID that starts with `prod_`, which can be obtained from the result of running the `stripe products crate` command or the `stripe products list` command.
+
+```sh
+curl http://localhost:8787/products/prod_xxxx
````
Add `$` at the first.

```markdown
$ curl http://localhost:8787/products/prod_xxxx
```

yusukebe commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
````diff
@@ -0,0 +1,847 @@
+This API requires the use of Workers AI and Vectorize. To use these resources from Workers, you need to create and bind the resources. First, let's create a Vectorize index with Wrangler using the command `wrangler vectorize create {index_name} --dimensions={number_of_dimensions} --metric={similarity_metric}`. The values for `dimensions` and `metric` depend on the type of [Text Embedding Model](/ ...
+
+```sh
+npx wrangler vectorize create stripe-products --dimensions=1024 --metric=cosine
````
Add `$` at the first.

```markdown
$ npx wrangler vectorize create stripe-products --dimensions=1024 --metric=cosine
```

yusukebe commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
````diff
@@ -0,0 +1,847 @@
+
+With this, we have established a mechanism to synchronize the product data with the database for recommendations. Use Stripe CLI commands to save some product data.
+
+```sh
````
With `sh`, we can copy the all lines:

https://github.com/cloudflare/cloudflare-docs/assets/10682/a427a87b-c688-44fd-b26c-5105c2dd0b15

So, how about using another format like a `bash`?

https://github.com/cloudflare/cloudflare-docs/assets/10682/a9b7e0d5-5d1d-4003-8277-c7384a499abf

yusukebe commented:
Hey @hideokamoto-stripe !

This is a great tutorial for Hono and Cloudflare AI stacks! I tried it and it works very well. I've left comments about formatting. Can you check them?

hideokamoto-stripe commented:
Thank you @yusukebe for reviewing my content!
I've updated this tutorial today.

Please check the diff that I fixed related to the format issues.
Thanks

yusukebe reviewed (approved):
LGTM!

dbenCF commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,851 @@
+
+Let's start implementing step-by-step.
+
+### Bind Workers AI and Vectorize to Workers
```
```suggestion
### Bind Workers AI and Vectorize to your Worker
```

dbenCF commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
````diff
@@ -0,0 +1,851 @@
+$ npm run dev -- --remote
+```
+
+### Create a Webhook API to Handle Product Registration Events
````
```suggestion
### Create a webhook API to handle product registration events
```

dbenCF commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,851 @@
+
+### Create a Webhook API to Handle Product Registration Events
+
+You can receive notifications about product registration and information via POST requests using Webhooks. Let's create an API that accepts POST requests. Open your `src/index.ts` file and add the following code:
```
```suggestion
You can receive notifications about product registration and information via POST requests using webhooks. Let's create an API that accepts POST requests. Open your `src/index.ts` file and add the following code:
```

dbenCF commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,851 @@
+
+This code implements an API that processes POST requests to the `/webhook` endpoint. The data sent by Stripe's Webhook events is included in the request body in JSON format. Therefore, we use `c.req.json()` to extract the data. There are multiple types of Webhook events that Stripe can send, so we added a conditional to only process events when a product is newly added, as indicated by the `type` ...
+
+### Adding Stripe's API Key to the Project
```
```suggestion
### Add Stripe's API Key to the project
```

dbenCF commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,851 @@
+
+### Adding Stripe's API Key to the Project
+
+When developing a Webhook API, you need to ensure that requests from unauthorized sources are rejected. To prevent unauthorized API requests from causing unintended behavior or operational confusion, you need a mechanism to verify the source of API requests. When integrating with Stripe, you can protect the API by generating a signing secret used for Webhook verification.
```
```suggestion
When developing a webhook API, you need to ensure that requests from unauthorized sources are rejected. To prevent unauthorized API requests from causing unintended behavior or operational confusion, you need a mechanism to verify the source of API requests. When integrating with Stripe, you can protect the API by generating a signing secret used for webhook verification.
```

dbenCF commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,851 @@
+8. Run `npm install stripe` to add the Stripe SDK to your application.
+9. Restart the `npm run dev -- --remote` command to import the API key into your application.
+
+Finally, modify the source code of `src/index.ts` as follows to ensure that the Webhook API cannot be used from sources other than your Stripe account.
```
```suggestion
Finally, modify the source code of `src/index.ts` as follows to ensure that the webhook API cannot be used from sources other than your Stripe account.
```

dbenCF reviewed (changes requested):
Absolutely amazing work! I left some suggestions for your review, let me know what you think of them.

kodster28 commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,851 @@
+
+You will also learn how to use the following Cloudflare products:
+
+- **Cloudflare Workers**: Execution environment for API applications
```
```suggestion
- [**Cloudflare Workers**](/workers/): Execution environment for API applications
```

kodster28 commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,851 @@
+You will also learn how to use the following Cloudflare products:
+
+- **Cloudflare Workers**: Execution environment for API applications
+- **Cloudflare Vectorize**: Vector DB used for related product searches
```
```suggestion
- [**Cloudflare Vectorize**](/vectorize/): Vector DB used for related product searches
```

kodster28 commented on content/workers-ai/tutorials/creating-a-recommendation-api.md:
```diff
@@ -0,0 +1,851 @@
+
+- **Cloudflare Workers**: Execution environment for API applications
+- **Cloudflare Vectorize**: Vector DB used for related product searches
+- **Cloudflare Workers AI**: Used for vectorizing data and generating recommendation texts
```
```suggestion
- [**Cloudflare Workers AI**](/workers-ai/): Used for vectorizing data and generating recommendation texts
```

hideokamoto-stripe commented:
Thank you @kodster28 for leaving such detailed feedback!
It's really helpful for me in improving this content.

I've now updated this PR to follow your feedback.
Best regards.
